subsys/bluetooth/host/hci_core.c

```diff
@@ -3277,6 +3277,13 @@ static int hci_init(void)
 {
 	int err;
 
+	if (bt_dev.drv->setup) {
+		err = bt_dev.drv->setup();
```

Comment:
```suggestion
		err = bt_dev.drv->setup(*bt_dev.drv->dev);
```

Reply:
This doesn't look very clean. So far we haven't exposed any struct device stuff to the host side, and none of the other HCI driver callbacks take this pointer.

Reply:
Or to put it another way, so far the host hasn't assumed that every HCI driver is built on top of some `struct device` abstraction. E.g. the userchan driver that native_posix uses doesn't have an underlying `struct device` concept (it simply has a file descriptor for read/write data)

Reply:
`struct device **dev` is a suggestion, `void ** userdata/param` or `struct bt_hci_driver *drv` are other alternative. It is very normal in driver model to have interface provided by a driver be supplied with an associated parent `dev/drv` or `userdata` parameter.

`h4_setup` as glue function to supply `h4_dev` is a hack, will lead to repeated such glue functions in every driver (ex. h5, spi, userchan) file that wish to implement a vendor `setup` interface.

Reply:
It's certainly not ideal if/when other driver need this kind of functionality, but I was assuming things could be refactored latest then. IIRC the Linux Bluetooth subsystem has a "minidriver" abstraction for vendor-specific setup, which can be reused by multiple HCI drivers.

As for Zephyr, the main reason the HCI driver API never evolved to have `struct device` is that it's a singleton (we don't support multiple controllers the way Linux does). Anyway, I'm not completely opposed to adding a context pointer, which could even be `struct device`, since that's fairly standard in Zephyr, but this needs to be made consistent throughout the HCI driver interface, i.e. it should really be the input to every callback, except `open()`. And then open() would return this pointer, so instead of `int open(void)` you'd have `struct device *open()`. (needs to be checked though what we loose with getting fail/success instead of a specific error code)

Reply:
>  open() would return this pointer, so instead of int open(void); you'd have struct device *open(void);

I thought about this, `open` to return `struct device *`, it could be ok as return. But we cannot expect `bt_enable` the caller of `open` to store and pass `struct device *dev` to `hci_init` and then thereafter to `setup`, will that be acceptable?

> I'm not completely opposed to adding a context pointer, which could even be struct device, since that's fairly standard in Zephyr, but this needs to be made consistent throughout the HCI driver interface, i.e. it should really be the input to every callback, except open()

Hope we can agree to this, add a `void *` context pointer, it will be generic enough so that drivers can pass data reference (`struct device *` or anything specific to that driver `setup`).

Reply:
hi @jhedberg, do we have agreement here?

Reply:
Not really. I was hoping to have time to discuss this with @cvinayak in our weekly Bluetooth call yesterday. One thing that’s noteworthy is that the original proposal would have been 100% backwards compatible without a need to go through the stable API change process. Now this has evolved to something that would immediately break any out-of-tree HCI driver.

Reply:
@jhedberg Adding `setup`, or adding both `setup` and a `data` now requires the stable API change process? `setup` was added into `struct bt_hci_driver` in the original proposal. With `CONFIG_BT_HCI_H4_SETUP=n`, `struct bt_hci_driver` will remain 100% backward compatible.

Reply:
Adding setup doesn’t require it, since omitted (not explicitly initialized) struct members are implicitly zeroed (or a NULL pointer in this case). Changing existing callbacks does cause breakage.

Reply:
This is no change to existing callbacks, the change requested was https://github.com/zephyrproject-rtos/zephyr/pull/41453#discussion_r783080377 or rather this https://github.com/zephyrproject-rtos/zephyr/pull/41453#discussion_r784758440

Reply:
There are several concerns with that proposal:

 - It’s inconsistent with the other callbacks (why shouldn’t they also get a data pointer?)
 - It makes it impossible to keep the driver struct in ROM if the data is not available at buildtime
 - It’s a bit odd to have to feed data back to the driver that the driver has already itself stored in its own struct
 
So it feels like this proposal is trying to forcefully mould the API to fit the the single use case at hand for a single driver, without looking at the overall driver API design and consistency.

Reply:
The `struct bt_hci_driver` originally has not followed driver model and was not designed for scalability in terms of adding enhancements; it does not support `struct device *` nor to have a custom parameter to the interfaces in it.

If we are willing to stick to this non-standard way of adding interface function to the driver and then having a glue function as workaround to pass parameters, I have no objections as this driver is already an exception, as long as the PR does not increase flash and RAM for samples that does not need `setup`.

Reply:
FWIW I’m all in favor of exploring transforming HCI drivers to a pure `struct device` based model, however that’s a different and much larger endeavor than what this PR is trying to address.

Reply:
@cvinayak I assume that the discussion regarding `struct device` conversion can move to a future PR, however there's one thing I wanted to point out here: even if/when the HCI driver interface is converted to `struct device`, the device pointer given to its callbacks will be that which represents the HCI driver instance itself. In the H4 driver case it will *not* be the device pointer of the UART. Since what's desired in this PR is a H4-specific setup callback (which is able to do UART-specific manipulations, such as getting & setting the baud rate) there would either way have to be a wrapper which passes the UART device pointer rather than the HCI driver device pointer.